File: Day07/Part02/day07_02.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays; 
import java.util.Map;
import java.util.HashMap;


class day07_02 {
    

    public static void main (String[] args) {

        System.out.println("Starting Day 07 Part 02...");

        HashMap<String, Program> programs = getPrograms("input.txt");

        System.out.println("Got programs...");

        // finding the base program by finding the program with the largest depth
        Integer biggestDepth = -1;
        Program baseProgram = null;

        for (Map.Entry<String, Program> entry : programs.entrySet()) {
            Integer currentDepth = entry.getValue().getDepth(programs);
            if (currentDepth > biggestDepth) {
                biggestDepth = currentDepth;
                baseProgram = entry.getValue();
            }
        };

        System.out.println("Base Program: " + baseProgram.getName());

        // next we need to start at the base and find all nodes that are unbalanced
        ArrayList<Program> unbalancedPrograms = new ArrayList<Program>();
        unbalancedPrograms.add(baseProgram);
        Program nextUnbalancedProgram = baseProgram.getNextUnbalanced(programs);
        while (nextUnbalancedProgram != null) {
            unbalancedPrograms.add(nextUnbalancedProgram);
            nextUnbalancedProgram = nextUnbalancedProgram.getNextUnbalanced(programs);
        }

        // get bad weight
        Integer badWeight = unbalancedPrograms.get(unbalancedPrograms.size() - 1).getTotalWeight(programs);
        System.out.println("Bad Weight: " + String.valueOf(badWeight));

        // get good weight
        Integer goodWeight = null;
        for (String s : unbalancedPrograms.get(unbalancedPrograms.size() - 2).getStackedPrograms()) {
            goodWeight = programs.get(s).getTotalWeight(programs);
            if (goodWeight != badWeight) {
                break;
            }
        }
        System.out.println("Target Weight: " + String.valueOf(goodWeight));

        // find difference between expected and actual weights, and add to the base weight
        Integer difference = goodWeight - badWeight;
        Integer baseWeight = unbalancedPrograms.get(unbalancedPrograms.size() - 1).getSpecificWeight();
        System.out.println("Expected Base Value: " + String.valueOf(baseWeight + difference));

    }


    public static HashMap<String, Program> getPrograms(String filename) {
        HashMap<String, Program> programs = new HashMap<String, Program>();
        try {
            BufferedReader br = new BufferedReader(new FileReader(filename));
            String next = br.readLine();
            while (next != null) {
                String[] stringProgram = next.split("\\W+");
                Program temp;
                if (stringProgram.length == 1) {
                    temp = new Program(stringProgram[0], null, null);
                } else if (stringProgram.length == 2) {
                    temp = new Program(stringProgram[0], Integer.valueOf(stringProgram[1]), null);
                } else if (stringProgram.length > 2) {
                    temp = new Program(stringProgram[0], Integer.valueOf(stringProgram[1]), Arrays.copyOfRange(stringProgram, 2, stringProgram.length));
                } else {
                    temp = null;
                }
                programs.put(temp.getName(), temp);
                next = br.readLine();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        if (programs.size() == 0) return null;
        return programs;
    }
    
}


class Program {

    private String mName;
    private Integer mWeight;
    private String[] mStackedPrograms;

    public Program(String name, Integer weight, String[] programs) {
        this.mName = name;
        this.mWeight = weight;
        this.mStackedPrograms = programs;
    }

    public String getName() {
        return mName;
    }

    public String[] getStackedPrograms() {
        return mStackedPrograms;
    }

    public void print() {
        System.out.println("Name: " + mName);
        System.out.println("Weight: " + mWeight);
        if (mStackedPrograms != null) {
            System.out.print("Stack: [ ");
            for (String s : mStackedPrograms) {
                System.out.print(s + " ");
            }
            System.out.print("]\n\n");
        } else {
            System.out.print("\n\n");
        }
    }

    public Integer getDepth(HashMap<String, Program> programs) {
        if (mStackedPrograms == null) {
            return 0;
        } else {
            Integer biggestDepth = 0;
            for (String s : mStackedPrograms) {
                Program p2 = programs.get(s);
                Integer depth = p2.getDepth(programs);
                biggestDepth = (depth > biggestDepth) ? depth : biggestDepth;
            }
            return biggestDepth + 1;
        }
    }

    public Integer getSpecificWeight() {
        return mWeight;
    }

    public Integer getTotalWeight(HashMap<String, Program> programs) {
        if (mStackedPrograms == null) {
            return this.getSpecificWeight();
        } else {
            Integer total = this.getSpecificWeight();
            for (String s : mStackedPrograms) {
                total += programs.get(s).getTotalWeight(programs);
            }
            return total;
        }
    }

    public Boolean isBalanced(HashMap<String, Program> programs) {
        if (mStackedPrograms == null) {
            return true;
        } else {
            Integer weight = null;
            Integer weightCompare = null;
            for (String s : mStackedPrograms) {
                if (weight == null) {
                    weight = programs.get(s).getTotalWeight(programs);
                } else {
                    weightCompare = programs.get(s).getTotalWeight(programs);
                    if (weight != weightCompare) {
                        return false;
                    }
                }
            }
            return true;
        }
    }

    public Program getNextUnbalanced(HashMap<String, Program> programs) {
        HashMap<Integer, Integer> counts = new HashMap<Integer, Integer>();
        Integer[] weights = new Integer[mStackedPrograms.length];
        for (int i = 0; i < mStackedPrograms.length; i++) {
            Integer weight = programs.get(mStackedPrograms[i]).getTotalWeight(programs);
            weights[i] = weight;
            if (counts.containsKey(weight)) {
                counts.put(weight, counts.get(weight)+1);
            } else {
                counts.put(weight, 1);
            }
        }
        for (int i = 0; i < weights.length; i++) {
            if (counts.get(weights[i]) == 1) {
                return programs.get(mStackedPrograms[i]);
            }
        }
        return null;
    }

}

File: Day07/Part02/day07_02.py
import re


class Program:

    def __init__(self, name, weight, programs):
        self.name = name
        self.weight = weight
        self.programs = programs

    @classmethod
    def fromList(cls, string_list):
        name = string_list[0]
        weight = int(string_list[1])
        programs = []
        for i in range(2, len(string_list)):
            if (string_list[i] != ''):
                programs.append(string_list[i])
        return cls(name, weight, programs)


def read_input(filename):
    f = open(filename, "r")
    programs = {}
    line = f.readline()
    while (line):
        split_line = re.split("\\W+", line)
        program = Program.fromList(split_line)
        programs[program.name] = program
        line = f.readline()
    f.close()
    return programs


def get_depth(program, program_dict):
    if (program.programs == []):
        return 0
    else:
        depth = 0
        for s in program.programs:
            d = get_depth(program_dict[s], program_dict)
            if (d > depth):
                depth = d
        return 1 + depth


def get_weight(program, program_dict):
    if (program.programs == []):
        return program.weight
    else:
        temp_weight = 0
        for s in program.programs:
            temp_weight += get_weight(program_dict[s], program_dict)
        return program.weight + temp_weight


def get_next_unbalanced(program, program_dict):
    weights = []
    for pstring in program.programs:
        weights.append(get_weight(program_dict[pstring], program_dict))
    for i,w in enumerate(weights):
        if weights.count(w) == 1:
            return program_dict[program.programs[i]]
    return None


if __name__ == "__main__":
    
    programs = read_input("input.txt")
    
    # find base as program with greatest depth
    depth = -1
    program = None
    for k,v in programs.items():
        d = get_depth(v, programs)
        if (d > depth):
            depth = d
            program = v
    
    # print result
    print("Base Program:", program.name)

    # find stack of unbalanced programs
    unbalanced_stack = [program]
    next_unknown = get_next_unbalanced(program, programs)

    while(next_unknown):
        unbalanced_stack.append(next_unknown)
        next_unknown = get_next_unbalanced(next_unknown, programs)

    # get the bad weight
    bad_weight = get_weight(unbalanced_stack[len(unbalanced_stack)-1], programs)
    print("Bad Weight:", bad_weight)

    # get the target weight
    target_weight = 0
    for pstring in unbalanced_stack[len(unbalanced_stack)-2].programs:
        target_weight = get_weight(programs[pstring], programs)
        if (target_weight != bad_weight):
            break
    print("Target Weight:", target_weight)

    # calculate required program weight to offset difference
    print("Required Program Weight:", unbalanced_stack[len(unbalanced_stack)-1].weight + (target_weight - bad_weight))
